src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,370 @@
+    (free_add_monoid.of (r, update f i m₁) + free_add_monoid.of (r, update f i m₂))
+    (free_add_monoid.of (r, update f i (m₁ + m₂)))
+| of_add_scalar : ∀ (r r' : R) (f : Π i, s i), eqv
+    (free_add_monoid.of (r, f) + free_add_monoid.of (r', f))
+    (free_add_monoid.of (r + r', f))
+| of_smul : ∀ (r : R) (f : Π i, s i) (i : ι) (r' : R), eqv
+    (free_add_monoid.of (r, update f i (r' • (f i))))
+    (free_add_monoid.of (r' * r, f))
+| add_comm : ∀ x y, eqv (x + y) (y + x)
+
+end pi_tensor_product
+
+variables (R) (s)
+
+/-- `pi_tensor_product R s` with `R` a commutative semiring and `s : ι → Type*` is the tensor
+  product of all the `s i`'s. This is denoted by `⨂[R] i, s i`. -/
+def pi_tensor_product : Type* :=
+(add_con_gen (pi_tensor_product.eqv R s)).quotient
+
+variables {R}
+
+/- This enables the notation `⨂[R] i : ι, s i` for the pi tensor product, given `s : ι → Type*`. -/
+localized "notation `⨂[`:100 R `] ` binders `, ` r:(scoped:67 f, pi_tensor_product R f) := r"
+  in tensor_product
+
+open_locale tensor_product
+
+namespace pi_tensor_product
+
+section module
+
+instance : add_comm_monoid (⨂[R] i, s i) :=
+{ add_comm := λ x y, add_con.induction_on₂ x y $ λ x y, quotient.sound' $
+    add_con_gen.rel.of _ _ $ eqv.add_comm _ _,
+  .. (add_con_gen (pi_tensor_product.eqv R s)).add_monoid }
+
+instance : inhabited (⨂[R] i, s i) := ⟨0⟩
+
+variables (R) {s}
+
+/-- `tprod_coeff R r f` with `r : R` and `f : Π i, s i` is the tensor product of the vectors `f i`
+over all `i : ι`, multiplied by the coefficient `r`. Note that this is meant as an auxiliary
+definition for this file alone, and that one should use `tprod` defined below for most purposes. -/
+def tprod_coeff (r : R) (f : Π i, s i) : ⨂[R] i, s i := add_con.mk' _ $ free_add_monoid.of (r, f)
+
+variables {R}
+
+lemma zero_tprod_coeff (f : Π i, s i) : tprod_coeff R 0 f = 0 :=
+quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_zero_scalar _
+
+lemma zero_tprod_coeff' (z : R) (f : Π i, s i) (i : ι) (hf: f i = 0) : tprod_coeff R z f = 0 :=
+quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_zero _ _ i hf
+
+lemma add_tprod_coeff (z : R) (f : Π i, s i) (i : ι) (m₁ m₂ : s i) :
+  tprod_coeff R z (update f i m₁) + tprod_coeff R z (update f i m₂) =
+    tprod_coeff R z (update f i (m₁ + m₂)) :=
+quotient.sound' $ add_con_gen.rel.of _ _ (eqv.of_add z f i m₁ m₂)
+
+lemma add_tprod_coeff' (z₁ z₂ : R) (f : Π i, s i) :
+  tprod_coeff R z₁ f + tprod_coeff R z₂ f = tprod_coeff R (z₁ + z₂) f :=
+quotient.sound' $ add_con_gen.rel.of _ _ (eqv.of_add_scalar z₁ z₂ f)
+
+lemma smul_tprod_coeff (z : R) (f : Π i, s i) (i : ι) (r : R) :
+  tprod_coeff R z (update f i (r • f i)) = tprod_coeff R (r * z) f :=
+quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_smul _ _ _ _
+
+/-- Construct an `add_monoid_hom` from `(⨂[R] i, s i)` to some space `F` from a function
+`φ : (R × Π i, s i) → F` with the appropriate properties. -/
+def lift_add_hom (φ : (R × Π i, s i) → F)
```

Comment:
I don't know it's a good idea, but it might be less work to make the signature of this
`(R →ₗ[R] multilinear_map R f F) → ((⨂[R] i, s i) →+ F)`, as then all your hypotheses are bundled.

Reply:
I'm not so sure it's worth it, since this is also something that probably shouldn't be used directly outside this file; `lift` is obviously much more useful.